UI: FileDropzone component to handle file list overwrite

zoltanbedi opened the pull request:
**What this PR does / why we need it**:

This pr adds the possibility to the FileDropzone component to overwrite the file list items. This is needed when we don't want to show the list items.

zoltanbedi committed:
UI: FileDropzone component to handle file list overwrite

zoltanbedi committed:
FileListItem: use type=button everywhere

tskarhed reviewed (changes requested):
Not sure if it should be handled in this PR, but the file list needs something to be read by screen readers for each item (Not sure if the file name works? Or the upload percentage?). 

I think I missed a couple of things last time around:

- The "Cancel" button could probably be changed to "Cancel upload". 
-  The upload icon could probably use a `aria-hidden`
- The default secondary text hould probably be a button
- ´aria-keyshortcuts´ could probably be used to inform screen reader that you can press enter to pick files

Just tell me if you want to do this in another PR, and I can approve this.

zoltanbedi committed:
Update packages/grafana-ui/src/components/FileDropzone/FileDropzone.story.tsx

Co-authored-by: Tobias Skarhed <1438972+tskarhed@users.noreply.github.com>

zoltanbedi committed:
FileListItem: add aria-hidden and change cancel text

zoltanbedi committed:
Merge branch 'zoltan/dropzone-update' of github.com:grafana/grafana into zoltan/dropzone-update

zoltanbedi commented:
> Not sure if it should be handled in this PR, but the file list needs something to be read by screen readers for each item (Not sure if the file name works? Or the upload percentage?).
> 
> I think I missed a couple of things last time around:
> 
> * The "Cancel" button could probably be changed to "Cancel upload".
> * The upload icon could probably use a `aria-hidden`
> * The default secondary text hould probably be a button
> * ´aria-keyshortcuts´ could probably be used to inform screen reader that you can press enter to pick files
> 
> Just tell me if you want to do this in another PR, and I can approve this.

Nah I think fixing these in this pr is good.

> * The "Cancel" button could probably be changed to "Cancel upload".

fixed
> * The upload icon could probably use a `aria-hidden`

fixed
> * The default secondary text should probably be a button

In the FileDropzone component? I think the current solution is fine like that. Should we check this with others?

> * ´aria-keyshortcuts´ could probably be used to inform screen reader that you can press enter to pick files

I'm not sure if we should add it. This is a browser behaviour not something custom that we added. What do you think should we still add it?

Clarity-89 commented on packages/grafana-ui/src/components/FileDropzone/FileListItem.test.tsx:
```diff
@@ -45,7 +45,7 @@ describe('The FileListItem component', () => {
     render(<FileListItem file={{ file: file({}), id: '1', error: null, progress: 6, abortUpload }} />);
 
-    fireEvent.click(screen.getByRole('button', { name: 'Cancel' }));
+    fireEvent.click(screen.getByRole('button', { name: 'Cancel upload' }));
```
If you want the tests to be a bit more robust, you can use regex to match text in this particular case (probably not a good idea with `*ByText` query as it may match unrelated elements.). This way you won't need to update test when there's some minor UI changes.
```suggestion
    fireEvent.click(screen.getByRole('button', { name: /cancel/i }));
```

Clarity-89 commented on packages/grafana-ui/src/components/FileDropzone/FileListItem.test.tsx:
```diff
@@ -30,7 +30,7 @@ describe('The FileListItem component', () => {
     render(<FileListItem file={{ file: file({}), id: '1', error: null, progress: 6 }} />);
 
-    expect(screen.queryByText('Cancel')).not.toBeInTheDocument();
+    expect(screen.queryByText('Cancel upload')).not.toBeInTheDocument();
```
```suggestion
    expect(screen.getByRole('button', { name: /cancel/i }))).not.toBeInTheDocument();
```
Even if it's `queryBy` query you can still use `*ByRole` here.

zoltanbedi replied:
actually this throws error because the element is not there. I check that it is not in the document here.

Clarity-89 replied:
Yeah, sorry, it was supposed to be `queryByRole` not `getByRole`.

Clarity-89 reviewed (approved):
Looks good!

zoltanbedi committed:
Update packages/grafana-ui/src/components/FileDropzone/FileListItem.test.tsx

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

tskarhed commented:
> > * The default secondary text should probably be a button
> 
> In the FileDropzone component? I think the current solution is fine like that. Should we check this with others?
> 
> > * ´aria-keyshortcuts´ could probably be used to inform screen reader that you can press enter to pick files
> 
> I'm not sure if we should add it. This is a browser behaviour not something custom that we added. What do you think should we still add it?

You're probably right. VoiceOver announces that the text is clickable, might be enough. And then you know to use Enter, so we can skip aria-keyshortcuts too.
